### PR TITLE
feat: support rescheduling table source & handle corner cases

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR handles some corner cases for rescheduling.

- Remove a non-existent parallel unit id.
- Add an occupied parallel unit id.
- Okay to reschedule materialize without downstream.
- Okay to migrate table source. (only to already existing compute node #5269)
- Do not panic on unsupported cases.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- #4243 
- #3750

### CODE REVIEW

File: src/source/src/table_v2.rs [hunk 1]
```diff
@@ -62,20 +61,34 @@ impl TableSourceV2 {
...
+                .iter()
+                .enumerate()
                 .choose(&mut rand::thread_rng())
-                .context("no available table reader in streaming source executors")?
-                .clone()
+                .context("no available table reader in streaming source executors")?;
+
+            // It's possible that the source executor is scaled in or migrated, so the channel
+            // is closed. In this case, we should remove the closed channel and retry.
+            if tx.is_closed() {
```
Comment (BugenZhao) [inline src/source/src/table_v2.rs:84]: This will be stale immediately. Should retry after really calling `send`.

**Review by shanicky — APPROVED** [review]
LSSTM!!!!!!

**Comment by BugenZhao** [comment]
> LSSTM!!!!!!

Could you please also help to test this PR with your test cases? ❤️

**Comment by yezizp2012** [comment]
If we reschedule source actors to newly joined CN, there will be anther thing need to be considering: the local source manager in new CN does not contains related source desc.

**Comment by BugenZhao** [comment]
> If we reschedule source actors to newly joined CN, there will be anther thing need to be considering: the local source manager in new CN does not contains related source desc.

Exactly. Do we have sync mechanism for sources now? 😢

**Comment by yezizp2012** [comment]
> > If we reschedule source actors to newly joined CN, there will be anther thing need to be considering: the local source manager in new CN does not contains related source desc.
> 
> Exactly. Do we have sync mechanism for sources now? 😢

Nope, only RPC calls. 🥵 Or we can let CN subscribe source info, it's a little bit costly but works.

**Comment by BugenZhao** [comment]
> > > If we reschedule source actors to newly joined CN, there will be anther thing need to be considering: the local source manager in new CN does not contains related source desc.
> > 
> > 
> > Exactly. Do we have sync mechanism for sources now? 😢
> 
> Nope, only RPC calls. 🥵 Or we can let CN subscribe source info, it's a little bit costly but works.

I'm not sure which pattern will be better for this, subscribing the source info, or totally removing the local source manager. As it's only required for table sources now, we may find a workaround way for this. 🤔

**Comment by yezizp2012** [comment]
> > > > If we reschedule source actors to newly joined CN, there will be anther thing need to be considering: the local source manager in new CN does not contains related source desc.
> > > 
> > > 
> > > Exactly. Do we have sync mechanism for sources now? 😢
> > 
> > 
> > Nope, only RPC calls. 🥵 Or we can let CN subscribe source info, it's a little bit costly but works.
> 
> I'm not sure which pattern will be better for this, subscribing the source info, or totally removing the local source manager. As it's only required for table sources now, we may find a workaround way for this. 🤔

Totally removing the local source manager seems not that easy currently, especially for table sources. 
IMO, if using subscribe mode, since `create_source` RPC in CN side is still needed for creating source ddl request (replacing with notification is not reliable as we discussed before), this will bring different behaviors between scheduling source actor and creating source. And CN should only subscribe source snapshot and not source info changes, which is quit weird.
If using RPC call here, we'd better make `create_source` able to accept duplicate sources. Because we might reschedule actors in same source multi times.

**Comment by BugenZhao** [comment]
I've opened an issue for the source manager in #5269.

**Pull request merged**